src/js/plugins/restrictions.js

```diff
@@ -0,0 +1,284 @@
+ripe.plugins = ripe.plugins || {};
+
+ripe.plugins.Restrictions = function(owner, restrictions, partsOptions, options) {
+    options = options || {};
+    this.token = options.token || ":";
+    this.owner = owner;
+    this.restrictions = restrictions;
+    this.restrictionsMap = this._buildRestrictionsMap(restrictions);
+    this.partsOptions = partsOptions;
+
+    // binds to the pre parts event so that the parts can be
+    // changed so that they comply with the product's restrictions
+    this.owner.bind("pre_parts", function(parts, newPart) {
+        // creates an array with the customization. If a new
+        // part is set it is added at the end so that it has
+        // priority when solving the restrictions
+        var partsOptions = ripe.clone(this.partsOptions);
+        var customization = [];
+        for (var name in parts) {
+            if (newPart !== undefined && newPart.name === name) {
+                continue;
+            }
+            var part = parts[name];
+            customization.push({
+                name: name,
+                material: part.material,
+                color: part.color
+            });
+        }
+        newPart !== undefined && customization.push(newPart);
+
+        // obtains the new parts and mutates the original
+        // parts map to apply the necessary changes
+        var newParts = this._solveRestrictions(
+            partsOptions,
+            this.restrictionsMap,
+            customization
+        );
+        for (var index = 0; index < newParts.length; index++) {
+            var newPart = newParts[index];
+            parts[newPart.name].material = newPart.material;
+            parts[newPart.name].color = newPart.color;
+        }
+    }.bind(this));
+};
+
+ripe.plugins.Restrictions.prototype._solveRestrictions = function(
+    availableParts,
+    restrictions,
+    customization,
+    solution
+) {
+    // if all the parts are set then a solution has been found
+    // and it is returned
+    solution = solution || [];
+    if (customization.length === 0 && this._isComplete(solution)) {
+        return solution;
+    }
+
+    // Retrieves a part from the customization and checks if it is
+    // being restricted by any of the validated parts. If there is
+    // no restriction then adds the part to the solution array and
+    // proceeds to the next part.
+    var newPart = customization.pop();
+    if (this._isRestricted(newPart, restrictions, solution) === false) {
+        solution.push(newPart);
+        return this._solveRestrictions(availableParts, restrictions, customization, solution);
+    }
+
+    // If the part is restricted then tries to retrieve an alternative option.
+    // If no alternative is found then an invalid state was reached and an empty
+    // solution is returned, otherwise tries to proceed with the alternative option.
+    else {

```

Comment:
the else is not really necessary since the if condition above returns and is easier to read without the else branch